File: trunk/subject-tracking/executor/cpp/PythonComponent.cpp
/******************************************************************************
 * NOTICE                                                                     *
 *                                                                            *
 * This software (or technical data) was produced for the U.S. Government     *
 * under contract, and is subject to the Rights in Data-General Clause        *
 * 52.227-14, Alt. IV (DEC 2007).                                             *
 *                                                                            *
 * Copyright 2024 The MITRE Corporation. All Rights Reserved.                 *
 ******************************************************************************/

/******************************************************************************
 * Copyright 2024 The MITRE Corporation                                       *
 *                                                                            *
 * Licensed under the Apache License, Version 2.0 (the "License");            *
 * you may not use this file except in compliance with the License.           *
 * You may obtain a copy of the License at                                    *
 *                                                                            *
 *    http://www.apache.org/licenses/LICENSE-2.0                              *
 *                                                                            *
 * Unless required by applicable law or agreed to in writing, software        *
 * distributed under the License is distributed on an "AS IS" BASIS,          *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
 * See the License for the specific language governing permissions and        *
 * limitations under the License.                                             *
 ******************************************************************************/

#include <exception>
#include <string>
#include <utility>

#include <pybind11/stl.h>

#include "detection.pb.h"
#include "PythonUtil.h"
#include "PythonScopedSignalHandler.h"

#include "PythonComponent.h"

namespace MPF::SUBJECT {

namespace {

    py::dict ToPython(const google::protobuf::Map<std::string, std::string>& map) {
        py::dict result;
        for (const auto& [key, value] : map) {
            result[py::str(key)] = value;
        }
        return result;
    }


    google::protobuf::Map<std::string, std::string> ConvertDict(const py::object& mapping) {
        google::protobuf::Map<std::string, std::string> result;
        for (const auto& pair : mapping.attr("items")()) {
            auto [key, value] = PythonUtil::ToStdPair<std::string, std::string>(pair);
            result[key] = std::move(value);
        }
        return result;
    }


    py::object ToPython(
            const mpf_buffers::ImageLocation& img_loc,
            const py::module_& detection_api) {
        return detection_api.attr("ImageLocation")(
            img_loc.x_left_upper(),
            img_loc.y_left_upper(),
            img_loc.width(),
            img_loc.height(),
            img_loc.confidence(),
            ToPython(img_loc.detection_properties())
        );
    }

    py::object ToPython(const mpf_buffers::VideoTrack& track, const py::module_& detection_api) {
        py::dict frame_locations;
        for (const auto& [frame, img_loc] : track.frame_locations()) {
            frame_locations[py::int_{frame}] = ToPython(img_loc, detection_api);
        }

        return detection_api.attr("VideoTrack")(
            track.start_frame(),
            track.stop_frame(),
            track.confidence(),
            frame_locations,
            ToPython(track.detection_properties())
        );
    }


    mpf_buffers::Entity ConvertEntity(py::handle py_entity) {
        mpf_buffers::Entity pb_entity;
        pb_entity.set_id(py::str(py_entity.attr("id")));
        pb_entity.set_score(1);
        auto& pb_tracks = *pb_entity.mutable_tracks();

        for (const auto& pair : py_entity.attr("tracks").attr("items")()) {
            auto [track_type, tracks] = PythonUtil::ToStdPair<std::string>(pair);
            auto& pb_track_id_list = pb_tracks[track_type];
            for (const auto& track_id : tracks) {
                pb_track_id_list.add_track_ids(track_id.cast<std::string>());
            }
        }
        *pb_entity.mutable_properties() = ConvertDict(py_entity.attr("properties"));
        return pb_entity;
    }

    mpf_buffers::Relationship ConvertRelationship(py::handle py_relationship) {
        mpf_buffers::Relationship pb_relationship;
        for (const auto& py_entity_id : py_relationship.attr("entities")) {
            pb_relationship.add_entities(py_entity_id.cast<py::str>());
        }

        auto& pb_frame_map = *pb_relationship.mutable_frames();
        for (const auto& py_media : py_relationship.attr("frames")) {
            auto& pb_frames = pb_frame_map[py_media.attr("id").cast<std::string>()];
            for (const auto& item : py_media.attr("frames")) {
                pb_frames.add_frames(item.cast<long>());
            }
        }
        *pb_relationship.mutable_properties() = ConvertDict(py_relationship.attr("properties"));
        return pb_relationship;
    }


    mpf_buffers::SubjectTrackingResult ResultsToCpp(py::handle py_results) {
        mpf_buffers::SubjectTrackingResult results;
        *results.mutable_properties() = ConvertDict(py_results.attr("properties"));

        auto& pb_entity_groups = *results.mutable_entity_groups();
        for (const auto& pair : py_results.attr("entities").attr("items")()) {
            auto [entity_type, entity_list] = PythonUtil::ToStdPair<std::string>(pair);
            auto& pb_entity_list = pb_entity_groups[entity_type];
            for (const auto& entity : entity_list) {
                *pb_entity_list.add_entities() = ConvertEntity(entity);
            }
        }

        auto& pb_relationship_groups = *results.mutable_relationship_groups();
        for (const auto& pair : py_results.attr("relationships").attr("items")()) {
            auto [relationship_type, relationship_list] = PythonUtil::ToStdPair<std::string>(pair);
            auto& pb_relationship_list = pb_relationship_groups[relationship_type];
            for (const auto& relationship : relationship_list) {
                *pb_relationship_list.add_relationships() = ConvertRelationship(relationship);
            }
        }
        return results;
    }

    mpf_buffers::SubjectTrackingResult CreateErrorResponse(const std::exception& exception) {
        mpf_buffers::SubjectTrackingResult result;
        result.add_errors(exception.what());
        return result;
    }
} // namespace

PythonComponent::PythonComponent(LoggerWrapper logger, std::string_view distribution_name)
        : logger_{std::move(logger)}
        , component_instance_{PythonUtil::LoadComponent(distribution_name)()} {
}

mpf_buffers::SubjectTrackingResult PythonComponent::GetSubjects(
        const mpf_buffers::SubjectTrackingJob& job) const {
    PythonScopedSignalHandler sig;
    auto py_job = ConvertSubjectTrackingJob(job);
    py::object py_results;
    try {
        py_results = component_instance_.attr("get_subjects")(py_job);
    }
    catch (const std::exception& e) {
        logger_.Error("The component threw an exception: ", e.what());
        return CreateErrorResponse(e);
    }
    return ResultsToCpp(py_results);
}

py::object PythonComponent::ConvertSubjectTrackingJob(
        const mpf_buffers::SubjectTrackingJob& subject_tracking_job) const {
    py::list image_jobs;
    for (const auto& image_job : subject_tracking_job.image_job_results()) {
        image_jobs.append(ConvertImageDetectionJob(image_job));
    }

    py::list video_jobs;
    for (const auto& video_job : subject_tracking_job.video_job_results()) {
        video_jobs.append(ConvertVideoDetectionJob(video_job));
    }
    return mpf_subject_api_.attr("SubjectTrackingJob")(
            subject_tracking_job.job_name(),
            ToPython(subject_tracking_job.job_properties()),
            video_jobs,
            image_jobs,
            py::list{},
            py::list{});
}

py::object PythonComponent::ConvertImageDetectionJob(
        const mpf_buffers::ImageDetectionJobResults& image_job) const {
    py::dict detection_results;
    for (const auto& [track_id, img_loc] : image_job.results()) {
        detection_results[py::str(track_id)] = ToPython(img_loc, mpf_component_api_);
    }
    const auto& detection_job = image_job.detection_job();
    return mpf_subject_api_.attr("ImageDetectionJobResults")(
            detection_job.data_uri(),
            detection_job.media_id(),
            detection_job.algorithm(),
            detection_job.track_type(),
            ToPython(detection_job.job_properties()),
            ToPython(detection_job.media_properties()),
            detection_results);
}


py::object PythonComponent::ConvertVideoDetectionJob(
        const mpf_buffers::VideoDetectionJobResults& video_job) const {
    py::dict detection_results;
    for (const auto& [track_id, track] : video_job.results()) {
        detection_results[py::str{track_id}] = ToPython(track, mpf_component_api_);
    }

    const auto& detection_job = video_job.detection_job();
    return mpf_subject_api_.attr("VideoDetectionJobResults")(
        detection_job.data_uri(),
        detection_job.media_id(),
        detection_job.algorithm(),
        detection_job.track_type(),
        ToPython(detection_job.job_properties()),
        ToPython(detection_job.media_properties()),
        detection_results
    );
}

} // namespace MPF::SUBJECT

File: trunk/subject-tracking/executor/python/mpf_subject_executor/python_component.py
#############################################################################
# NOTICE                                                                    #
#                                                                           #
# This software (or technical data) was produced for the U.S. Government    #
# under contract, and is subject to the Rights in Data-General Clause       #
# 52.227-14, Alt. IV (DEC 2007).                                            #
#                                                                           #
# Copyright 2024 The MITRE Corporation. All Rights Reserved.                #
#############################################################################

#############################################################################
# Copyright 2024 The MITRE Corporation                                      #
#                                                                           #
# Licensed under the Apache License, Version 2.0 (the "License");           #
# you may not use this file except in compliance with the License.          #
# You may obtain a copy of the License at                                   #
#                                                                           #
#    http://www.apache.org/licenses/LICENSE-2.0                             #
#                                                                           #
# Unless required by applicable law or agreed to in writing, software       #
# distributed under the License is distributed on an "AS IS" BASIS,         #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  #
# See the License for the specific language governing permissions and       #
# limitations under the License.                                            #
#############################################################################

import importlib.metadata
from typing import Type

import mpf_component_api as mpf
import mpf_subject_api as mpf_sub

from . import executor_util as util
from .logger_wrapper import LoggerWrapper
from .mpf_proto import detection_pb2 as pb_det
from .mpf_proto import subject_pb2 as pb_sub


class PythonComponent:
    def __init__(self, logger: LoggerWrapper, distribution_name: str):
        self._logger = logger
        self._component_instance = self._load_component_cls(distribution_name)()


    def get_subjects(self, pb_job: pb_sub.SubjectTrackingJob) -> pb_sub.SubjectTrackingResult:
        job = ProtobufToSubjectJobConverter.convert(pb_job)
        try:
            results = self._component_instance.get_subjects(job)
        except Exception as e:
            self._logger.exception(
                'Creating error response because the component threw an exception: ', e)
            return pb_sub.SubjectTrackingResult(errors=[str(e)])
        return SubjectResultToProtobufConverter.convert(results)


    @staticmethod
    def _load_component_cls(distribution_name: str) -> Type[mpf_sub.SubjectTrackingComponent]:
        try:
            distribution = importlib.metadata.distribution(distribution_name)
            if len(distribution.entry_points) == 1:
                return next(iter(distribution.entry_points)).load()
            group_matches = None
            for entry_point in distribution.entry_points:
                if entry_point.group != 'mpf.exported_component':
                    continue
                if entry_point.name == 'component':
                    return entry_point.load()
                group_matches = entry_point
            if group_matches:
                # An entry point in the "mpf.exported_component" group was found, but the
                # left-hand side of the '=' was something else. For example
                # 'mpf.exported_component': 'MyComponentClass = my_module:MyComponentClass'
                # We really only care about the entry point group, since we don't do anything
                # with entry point name.
                return group_matches.load()
        except importlib.metadata.PackageNotFoundError as e:
            raise util.ComponentLoadError(
                    f'Could not find package named {distribution_name}') from e
        except Exception as e:
            raise util.ComponentLoadError(f'Failed to load component due to: {e}') from e
        raise util.ComponentLoadError('Component entrypoint not present.')




class ProtobufToSubjectJobConverter:
    @classmethod
    def convert(cls, pb_job: pb_sub.SubjectTrackingJob) -> mpf_sub.SubjectTrackingJob:
        video_jobs = [cls._convert_video_job(j) for j in pb_job.video_job_results]
        image_jobs = [cls._convert_image_job(j) for j in pb_job.image_job_results]
        audio_jobs = [cls._convert_audio_job(j) for j in pb_job.audio_job_results]
        generic_jobs = [cls._convert_generic_job(j) for j in pb_job.generic_job_results]
        return mpf_sub.SubjectTrackingJob(
            pb_job.job_name,
            dict(pb_job.job_properties),
            video_jobs,
            image_jobs,
            audio_jobs,
            generic_jobs)


    @classmethod
    def _convert_video_job(
            cls, pb_vid_job: pb_sub.VideoDetectionJobResults) -> mpf_sub.VideoDetectionJobResults:
        detection_results = {
                mpf_sub.TrackId(t_id): cls._convert_video_track(track)
                for t_id, track in  pb_vid_job.results.items()}
        detection_job = pb_vid_job.detection_job
        return mpf_sub.VideoDetectionJobResults(
            detection_job.data_uri,
            mpf_sub.MediaId(detection_job.media_id),
            detection_job.algorithm,
            mpf_sub.DetectionComponentType(detection_job.track_type),
            dict(detection_job.job_properties),
            dict(detection_job.media_properties),
            detection_results)


    @classmethod
    def _convert_image_job(
            cls, pb_img_job: pb_sub.ImageDetectionJobResults) -> mpf_sub.ImageDetectionJobResults:
        detection_results = {
                mpf_sub.TrackId(t_id): cls._convert_image_detection(img_loc)
                for t_id, img_loc in pb_img_job.results.items()}

        detection_job = pb_img_job.detection_job
        return mpf_sub.ImageDetectionJobResults(
            detection_job.data_uri,
            mpf_sub.MediaId(detection_job.media_id),
            detection_job.algorithm,
            mpf_sub.DetectionComponentType(detection_job.track_type),
            dict(detection_job.job_properties),
            dict(detection_job.media_properties),
            detection_results)

    @classmethod
    def _convert_video_track(cls, track: pb_det.VideoTrack) -> mpf.VideoTrack:
        frame_locations = {
            f: cls._convert_image_detection(img_loc)
            for f, img_loc in track.frame_locations.items()}
        return mpf.VideoTrack(
            track.start_frame,
            track.stop_frame,
            track.confidence,
            frame_locations,
            dict(track.detection_properties))

    @staticmethod
    def _convert_image_detection(pb_img_detection: pb_det.ImageLocation) -> mpf.ImageLocation:
        return mpf.ImageLocation(
            pb_img_detection.x_left_upper,
            pb_img_detection.y_left_upper,
            pb_img_detection.width,
            pb_img_detection.height,
            pb_img_detection.confidence,
            dict(pb_img_detection.detection_properties))


    @classmethod
    def _convert_audio_job(
            cls, pb_job: pb_sub.AudioDetectionJobResults) -> mpf_sub.AudioDetectionJobResults:
        detection_results = {
            mpf_sub.TrackId(t_id): cls._convert_audio_track(track)
            for t_id, track in pb_job.results.items()
        }
        detection_job = pb_job.detection_job
        return mpf_sub.AudioDetectionJobResults(
            detection_job.data_uri,
            mpf_sub.MediaId(detection_job.media_id),
            detection_job.algorithm,
            mpf_sub.DetectionComponentType(detection_job.track_type),
            dict(detection_job.job_properties),
            dict(detection_job.media_properties),
            detection_results)

    @staticmethod
    def _convert_audio_track(pb_track: pb_det.AudioTrack) -> mpf.AudioTrack:
        return mpf.AudioTrack(
            pb_track.start_time,
            pb_track.stop_time,
            pb_track.confidence,
            dict(pb_track.detection_properties))


    @classmethod
    def _convert_generic_job(
            cls, pb_job: pb_sub.GenericDetectionJobResults) -> mpf_sub.GenericDetectionJobResults:
        detection_results = {
            mpf_sub.TrackId(t_id): cls._convert_generic_track(track)
            for t_id, track in pb_job.results.items()
        }
        detection_job = pb_job.detection_job
        return mpf_sub.GenericDetectionJobResults(
            detection_job.data_uri,
            mpf_sub.MediaId(detection_job.media_id),
            detection_job.algorithm,
            mpf_sub.DetectionComponentType(detection_job.track_type),
            dict(detection_job.job_properties),
            dict(detection_job.media_properties),
            detection_results)


    @staticmethod
    def _convert_generic_track(pb_track: pb_det.GenericTrack) -> mpf.GenericTrack:
        return mpf.GenericTrack(pb_track.confidence, dict(pb_track.detection_properties))



class SubjectResultToProtobufConverter:

    @classmethod
    def convert(cls, results: mpf_sub.SubjectTrackingResults) -> pb_sub.SubjectTrackingResult:
        pb_entity_groups = {}
        for type_, entities in results.entities.items():
            pb_entity_groups[type_] = pb_sub.EntityList(
                    entities=[cls._convert_entity(e) for e in entities])

        pb_rel_groups = {}
        for type_, rels in results.relationships.items():
            pb_rel_groups[type_] = pb_sub.RelationshipList(
                    relationships=[cls._convert_relationship(r) for r in rels])

        return pb_sub.SubjectTrackingResult(
                entity_groups=pb_entity_groups,
                relationship_groups=pb_rel_groups,
                properties=results.properties,
                errors=())


    @staticmethod
    def _convert_entity(entity: mpf_sub.Entity) -> pb_sub.Entity:
        pb_tracks = {
                str(type_): pb_sub.TrackIdList(track_ids=ids)
                for type_, ids in entity.tracks.items()}
        return pb_sub.Entity(
            id=str(entity.id),
            score=entity.score,
            tracks=pb_tracks,
            properties=entity.properties
        )

    @staticmethod
    def _convert_relationship(relationship: mpf_sub.Relationship) -> pb_sub.Relationship:
        return pb_sub.Relationship(
            entities=[str(e_id) for e_id in relationship.entities],
            frames={m.id: pb_sub.FrameList(frames=m.frames) for m in relationship.frames},
            properties=relationship.properties
        )
